### ISSUE TITLE
Long compilation time

### ISSUE THREAD

**Issue by mpatel31415** [Lightning-AI/lightning-thunder#229]
## 🐛 Bug

With newest version of Docker image (tested on 2024-04-05 and later) during training with thunder.jit with default executors now compilation takes more time comparing to image with tag pjnl-pipeline-13351802-amd64 and to `torch.compile`.

The results are mostly visible for :
pythia-410m with bs = 1,  8 gpus, 1 node and training with FSPD + zero2 (88% increase in time of time of compilation)
but can be observed for other models e.g. pythia-6.9b (71% increase) or dolly-v2-7b (71%) with FSPD + zero3. 

![image](https://github.com/Lightning-AI/lightning-thunder/assets/149149379/e779a29c-5570-45fb-b8b4-c74d2e9343d6)

Approximate compilation time is calculated as overhead in iteration time observed in the first half of measured iterations, comparing to iteration time after warmup. 

In details: Assuming we run 2*N training iterations and we have iter_times[i] for them:
iter_time_after_warmup = mean( iter_times[N:-1]) 
approx_compile_time = sum[iter_times[i] - iter_time_after_warmup for i in range(0, N)]

### To Reproduce
 As example I can provide the code below, but the results presented in the image were gathered from slurm jobs on 8 GPUs. 
To gather more data I used internal cluster and tools, you can contact me and I'll share the details on how to reproduce them.

**Thunder:**

```
docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864  --pull always -it INTERNAL_ADDRESS:5005/dl/pytorch/update-scripts:pjnl-latest

# inside:
torchrun --nproc-per-node=4 /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py --model_name dolly-v2-7b --compile thunder --distributed_mode fsdp --shard_mode zero2
```
Results:
iter 0: loss 11.0000, iter time: 290983.09ms, t: 2048
iter 1: loss 8.4375, iter time: 1324.23ms, t: 2048

**Inductor:**
For newest image (tag pjnl-20240418) we get an error:
An error occurred: InternalTorchDynamoError – 'NoneType' object has no attribute '__code__'

For older image:
```
docker run --gpus all --ipc=host --ulimit memlock=-1 --ulimit stack=67108864 -it INTERNAL_ADDRESS:5005/dl/pytorch/update-scripts:pjnl-20240315 

# inside
python -m pip install jsonargparse
torchrun --nproc-per-node=4 /opt/pytorch/lightning-thunder/thunder/benchmarks/benchmark_litgpt.py --model_name dolly-v2-7b --compile inductor --distributed_mode fsdp --shard_mode zero2
```

Results:
iter 0: loss 11.0000, iter time: 49965.17ms, t: 2048
iter 1: loss 8.4375, iter time: 1743.59ms, t: 2048

**Thunder in old container**
(code is the same as for Inductor, except --compile argument)

Results:
iter 0: loss 11.0000, iter time: 177376.48ms, t: 2048

Before each testing each compilation method I restarted the container.

### Expected behavior

Ideally compilation time should not increase from with development and should be similar to cases when using inductor.

Environment
As in the Docker image. The reproduction examples come from 4xH100 GPUs.

![image](https://github.com/Lightning-AI/lightning-thunder/assets/149149379/fde95719-1a12-4535-af87-bd7fff7c0cef)

### Additional context
Similar behaviour is visible for other models, batch sizes and  # gpus.

cc @tfogal

**Comment by mruberry** [comment]
triage review:

- likely related to the switch from `compile()` to `jit()`
- a more general tracker on interpreter performance may be a useful issue
- it would be interesting to profile the interpreter and identify opportunities to improve its performance (flamegraph/heatmap)
- @riccardofelluga — is this something you'd be interested in?

**Comment by IvanYashchuk** [comment]
The slowdown is due to this https://github.com/NVIDIA/Fuser/commit/36c9cd01da695b5c97323d0008edc7fe8ffb80b9 commit. @samnordmann, @naoyam do you know what could be causing the slowdown in the initial iteration time? What kind of information would be useful for you to help debug this problem?

**Comment by riccardofelluga** [comment]
## Info on the issue

This does't seem to be an issue relate to the switch between `compile()` and `jit()`. Furthermore, the flamegraph seems to be in favor of the `jit()` transition preformance wise as there are fewer flat peaks. In the following pictures you can see the comparison between the images `20240315` and `20240427`:

![flame_315](https://github.com/Lightning-AI/lightning-thunder/assets/11768013/47cb5351-3aaf-4230-87d7-40228f87ce7c)

![flame_427](https://github.com/Lightning-AI/lightning-thunder/assets/11768013/5cd785d9-9266-4e6f-8e72-624b872ad679)

However, this issue seems to be related to nvfuser as the profiling with nsys shows. In particular the issue seems to be localized on the time spent by nvFuser to find a fusion for the backward pass when there are LayerNorm and Linear layers present(as in the the `dolly-v2-7B` blocks). 

![32_blocks_old](https://github.com/Lightning-AI/lightning-thunder/assets/11768013/782b8d35-bbb1-449a-831b-f72d3bcfc454)

![32_blocks_new](https://github.com/Lightning-AI/lightning-thunder/assets/11768013/59521294-0b3a-4d20-9c45-d459e2b28fc7)

One might say that Thunder is doing something in the transforms that makes it very hard for nvFuser to find a fusion, however after further inspection of the traces there does not seem to be meaningful difference between them.

Opening an issue on nvFuser to investigate this some more.

## Repro

This numbers can be reproduced with the following:

``` python
import torch
import torch.nn as nn
import thunder
import time

class MockDolly(nn.Module):
    def __init__(self, n_mock_blocks) -> None:
        super().__init__()
        self.layers = nn.Sequential(
            *[torch.nn.LayerNorm((4096,), eps=1e-05, elementwise_affine=True),
            nn.Linear(4096, 4096),
            torch.nn.LayerNorm((4096,), eps=1e-05, elementwise_affine=True),
            nn.Linear(4096, 4096)]*n_mock_blocks,
            torch.nn.LayerNorm((4096,), eps=1e-05, elementwise_affine=True)
    )
    def forward(self, x):
        return self.layers(x)

a = torch.randn((1, 2048, 4096), device="cuda:0", dtype=torch.bfloat16)
model = MockDolly(2)

model.to(dtype=torch.bfloat16)
model.to(device="cuda:0")

model = thunder.jit(model)
st = time.perf_counter_ns()
out = model(a)
torch.cuda.synchronize()
print(time.perf_counter_ns() - st)

st = time.perf_counter_ns()
out.sum().backward()
torch.cuda.synchronize()
print(time.perf_counter_ns() - st)
```

**Comment by samnordmann** [comment]
> The slowdown is due to this [NVIDIA/Fuser@36c9cd0](https://github.com/NVIDIA/Fuser/commit/36c9cd01da695b5c97323d0008edc7fe8ffb80b9) commit. @samnordmann, @naoyam do you know what could be causing the slowdown in the initial iteration time? What kind of information would be useful for you to help debug this problem? (The slowdown is still present with the latest nvFuser)

If the slowdown comes indeed from this commit, my main guess would be to look at [this line](https://github.com/NVIDIA/Fuser/blob/729f36c6fd0b86aa9d09a89ff3e406978cda13c7/csrc/scheduler/utils.cpp#L1071) which triggers the creation of a `ComputeAtMap` down the line [here](https://github.com/NVIDIA/Fuser/blob/729f36c6fd0b86aa9d09a89ff3e406978cda13c7/csrc/multidevice/utils.h#L37) which is relatively costly. Does it make sense?

Before coming up with a proper fix, @IvanYashchuk could try commenting out this line and let me know whether the slow down persists?

**Comment by IvanYashchuk** [comment]
I can confirm that commenting out this line fixes the problem!

**Comment by riccardofelluga** [comment]
I can confirm that also with the previously posted repro, ran on single gpu the problem can be fixed by commenting that statement. For the backward pass we go from 8s to 5s on on an A100.

**Comment by samnordmann** [comment]
I'll push a fix in a couple of weeks. In the meantime, if urgent please feel free to revert the commit

**Comment by apaz-cli** [comment]
Also notable is that we haven't put any effort into optimizing the interpreter thus far. We know that it's slow, and we know that slowdown is at least in part due to the crazy number of nested function calls, as well as the necessity of using exceptions as control flow (because that's what cpython does). I can take a look at it when I return to the project.

**Comment by csarofeen** [comment]
CC @wujingyue

**Comment by wujingyue** [comment]
I ran the reproducer in https://github.com/Lightning-AI/lightning-thunder/issues/229#issuecomment-2098252103. 

Before https://github.com/NVIDIA/Fuser/pull/2234, which is now merged: 
```
me @ swagtron : stage | /opt/pytorch/lightning-thunder (main*)
$ python repro.py
1855177022
5052094196
```

After https://github.com/NVIDIA/Fuser/pull/2234:
```
me @ swagtron : stage | /opt/pytorch/lightning-thunder (main*)
$ python repro.py
1767807499
3390632109
```

There's indeed some improvement, but is that good enough to close this bug? @riccardofelluga

**Comment by riccardofelluga** [comment]
I've briefly tested it without profiling and it looks good to me!
On the original problem we went from before `05cba8c`:
``` shell
iter 0: loss 11.0000, iter time: 214517.46ms, t: 2048
iter 1: loss 8.4375, iter time: 384.82ms, t: 2048
iter 2: loss 5.3750, iter time: 271.40ms, t: 2048
```
to after `05cba8c`:
``` shell
iter 0: loss 11.0000, iter time: 129401.06ms, t: 2048
iter 1: loss 8.4375, iter time: 270.39ms, t: 2048
iter 2: loss 5.3750, iter time: 271.10ms, t: 2048
```
Thanks everybody!!

Closing for now, however we might still want to inspect interpreter performance in the future.

**Comment by apaz-cli** [comment]
@riccardofelluga Out of curiosity, what tool did you use to generate the flamegraph? I've been looking around for a better python flamegraph tool.
